File: 2019/batch_01/l005_Heap/heapSort.java
public class heapSort {
    public static void main(String[] args) {
        int[] arr = { 10, 20, 30, -2, -3, -4, 5, 6, 7, 8, 9, 22, 11, 13 };
        heapsort(arr, true);
    }

    public static void heapsort(int[] arr, boolean isMax) {
        for (int i = arr.length - 1; i >= 0; i--) {
            downHeapify(arr, i, arr.length - 1, isMax);
        }

        for (int i = 0; i < arr.length; i++) {
            for (int ele : arr) {
                System.out.print(ele + " ");
            }

            System.out.println();

            swap(arr, 0, arr.length - i - 1);
            downHeapify(arr, 0, arr.length - 1 - i, isMax);
        }

    }

    public static void downHeapify(int[] arr, int idx, int n, boolean isMax) {
        int maxidx = idx;
        int lci = 2 * idx + 1;
        int rci = 2 * idx + 2;

        if (lci < n && compareTo(arr[lci], arr[maxidx], isMax) > 0) {
            maxidx = lci;
        }

        if (rci < n && compareTo(arr[rci], arr[maxidx], isMax) > 0) {
            maxidx = rci;
        }

        if (maxidx != idx) {
            swap(arr, maxidx, idx);
            downHeapify(arr, maxidx, n, isMax);
        }
    }

    public static int compareTo(int val1, int val2, boolean isMax) {
        if (isMax) {
            return (val1 - val2);
        } else {
            return val2 - val1;
        }
    }

    public static void swap(int[] arr, int x, int y) {
        int temp = arr[x];
        arr[x] = arr[y];
        arr[y] = temp;
    }

}

File: 2019/levelUp302/Lecture_007_HashMapAndHeap/HeapConstruction/HeapSort.cpp
#include <iostream>
#include <vector>

using namespace std;

void downHeapify(vector<int> &arr, int idx, int n)
{
    int lci = 2 * idx + 1;
    int rci = 2 * idx + 2;
    int maxIdx = idx;

    if (lci <= n && arr[lci] > arr[maxIdx])
        maxIdx = lci;
    if (rci <= n && arr[rci] > arr[maxIdx])
        maxIdx = rci;

    if (maxIdx != idx)
    {
        swap(arr[idx], arr[maxIdx]);
        downHeapify(arr, maxIdx, n);
    }
}

void heapSort(vector<int> &arr)
{
    int n = arr.size() - 1;
    for (int i = n; i >= 0; i--)
        downHeapify(arr, i, n);

    int idx = 0;
    while (n != 0)
    {
        swap(arr[0], arr[n--]);
        downHeapify(arr, 0, n);
    }

    for (int ele : arr)
        cout << ele << " ";
}

void solve()
{
    vector<int> arr{10, 20, 30, -2, -3, -4, 5, 6, 7, 8, 9, 22, 11, 13};
    heapSort(arr);
}

int main()
{
    solve();
    return 0;
}
